File: Configs/.local/lib/hyde/hypr.altab.py
from __future__ import annotations

import json
import os
import shutil
import socket
import subprocess
import sys
from typing import Any, List


DEBUG = "--debug" in sys.argv[1:] or os.environ.get("HYPR_ALTAB_DEBUG", "0") == "1"
NOTIFY = os.environ.get("HYPR_ALTAB_NOTIFY", "1") != "0"
if "--notify" in sys.argv[1:]:
    NOTIFY = True
if "--no-notify" in sys.argv[1:]:
    NOTIFY = False
PREV = "--prev" in sys.argv[1:]
STATE_DIR = os.path.join(os.environ.get("XDG_RUNTIME_DIR", "/tmp"), "hypr-altab")
STATE_FILE = os.path.join(STATE_DIR, "state")
PREVIEW_DIR = os.path.join(STATE_DIR, "previews")
_PREVIEW_NEXT_ENV = os.environ.get("HYPR_ALTAB_PREVIEW_NEXT")
PREVIEW_NEXT = int(_PREVIEW_NEXT_ENV) if _PREVIEW_NEXT_ENV is not None else -1
HIS = os.environ.get("HYPRLAND_INSTANCE_SIGNATURE", "")
XDG_RUNTIME_DIR = os.environ.get("XDG_RUNTIME_DIR", "/tmp")
SOCKET_PATH = os.path.join(XDG_RUNTIME_DIR, "hypr", HIS, ".socket.sock")


def log(msg: str) -> None:
    if DEBUG:
        print(f"[hypr-altab] {msg}", file=sys.stderr, flush=True)


def run(cmd: List[str]) -> subprocess.CompletedProcess[str]:
    log(f"run: {' '.join(cmd)}")
    proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    if DEBUG and (proc.stdout or proc.stderr):
        if proc.stdout:
            log(f"stdout: {proc.stdout.strip()}")
        if proc.stderr:
            log(f"stderr: {proc.stderr.strip()}")
    return proc


def hyprctl_raw(command: str) -> str:
    if not HIS:
        log("HYPRLAND_INSTANCE_SIGNATURE not set")
        return ""
    try:
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.connect(SOCKET_PATH)
            sock.sendall(command.encode("utf-8"))
            sock.shutdown(socket.SHUT_WR)
            data = sock.recv(10_000_000)
            return data.decode("utf-8", errors="replace")
    except OSError as exc:
        log(f"socket error: {exc}")
        return ""


def hyprctl_json(command: str) -> Any:
    raw = hyprctl_raw(command)
    if not raw:
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return None


def focus_addr(addr: str) -> None:
    if not addr:
        log("focus_addr called with empty address")
        return
    hyprctl_raw(f"dispatch focuswindow address:{addr}")


def preview_path(addr: str) -> str:
    safe = addr.replace("0x", "").replace(":", "")
    return os.path.join(PREVIEW_DIR, f"{safe}.png")


def capture_preview(addr: str) -> None:
    if not addr:
        return
    if not shutil.which("grimblast"):
        return
    os.makedirs(PREVIEW_DIR, exist_ok=True)
    path = preview_path(addr)
    run(["grimblast", "save", "active", path])


def read_state() -> dict[str, Any]:
    if not os.path.exists(STATE_FILE):
        return {}
    try:
        with open(STATE_FILE, "r", encoding="utf-8") as handle:
            return json.load(handle)
    except (OSError, json.JSONDecodeError):
        return {}


def build_notify(addr: str) -> tuple[str, str, str | None]:
    clients = hyprctl_json("j/clients")
    if not isinstance(clients, list):
        return "", "", None
    match = next((c for c in clients if c.get("address") == addr), None)
    if not match:
        return "", "", None
    title = match.get("title") or "(untitled)"
    klass = match.get("class") or "unknown"
    ws = (match.get("workspace") or {}).get("name") or "?"
    body = f"{title}\n{klass}  •  {ws}"
    icon = preview_path(addr)
    return klass, body, icon if os.path.exists(icon) else None


def notify_preview_pair(current_addr: str, next_addrs: List[str]) -> None:
    if not shutil.which("notify-send"):
        return
    for offset, addr in enumerate(next_addrs, start=1):
        title, body, icon = build_notify(addr)
        if title:
            cmd = ["notify-send", title, body, "-t", "1500", "-r", str(6 + offset)]
            if icon:
                cmd.extend(["-i", icon])
            run(cmd)
    title, body, icon = build_notify(current_addr)
    if title:
        cmd = ["notify-send", title, body, "-t", "2000", "-r", "6"]
        if icon:
            cmd.extend(["-i", icon])
        run(cmd)


def write_state(state: dict[str, Any]) -> None:
    os.makedirs(STATE_DIR, exist_ok=True)
    with open(STATE_FILE, "w", encoding="utf-8") as handle:
        json.dump(state, handle)


def clear_state() -> None:
    try:
        os.remove(STATE_FILE)
    except OSError:
        pass


def get_active_address() -> str:
    active = hyprctl_json("j/activewindow")
    if isinstance(active, dict):
        return active.get("address") or ""
    return ""


def build_history() -> List[str]:
    clients = hyprctl_json("j/clients")
    if not isinstance(clients, list):
        return []
    filtered = [
        c
        for c in clients
        if c.get("mapped") is True and ((c.get("hidden") is False) or (c.get("grouped") or []))
    ]
    ordered = sorted(
        [c for c in filtered if isinstance(c.get("focusHistoryID"), int)],
        key=lambda c: c.get("focusHistoryID"),
    )
    return [c.get("address") for c in ordered if c.get("address")]


def main() -> int:
    if "--cancel" in sys.argv[1:]:
        clear_state()
        return 0
    if "--apply" in sys.argv[1:]:
        state = read_state()
        addr = state.get("addr", "")
        log(f"apply: target={addr}")
        if addr:
            focus_addr(addr)
            capture_preview(addr)
        clear_state()
        return 0

    state = read_state()
    history = state.get("list") if isinstance(state.get("list"), list) else None
    if not history:
        history = build_history()
        if len(history) < 2:
            log("not enough windows in focus history")
            return 0

    if PREV:
        idx = int(state.get("idx", 0)) - 1
        if idx <= 0:
            idx = len(history) - 1
    else:
        idx = int(state.get("idx", 0)) + 1
        if idx >= len(history):
            idx = 1

    target = history[idx]
    next_targets: List[str] = []
    preview_count = PREVIEW_NEXT if PREVIEW_NEXT > 0 else max(1, len(history) - 1)
    if len(history) > 1 and preview_count > 0:
        for step in range(1, preview_count + 1):
            if PREV:
                next_idx = idx - step
                if next_idx <= 0:
                    next_idx = len(history) - 1 - ((step - idx) % max(1, len(history) - 1))
            else:
                next_idx = idx + step
                if next_idx >= len(history):
                    next_idx = 1 + (next_idx - len(history)) % max(1, len(history) - 1)
            next_targets.append(history[next_idx])
    log(f"preview: idx={idx} target={target} count={len(history)} active={get_active_address()}")
    write_state({"idx": idx, "addr": target, "list": history})
    if NOTIFY:
        notify_preview_pair(target, next_targets)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Configs/.local/lib/hyde/hypr.altab.lua
local lib_dir = os.getenv("LIB_DIR") or "."
local v_require = dofile(lib_dir .. "/hyde/luautils/require.lua")

local socket = require("socket.unix")
local json = require("dkjson")
local argparse = require "argparse"

local parser = argparse("hyde-shell hypr.altab", "Hyprland Alt-Tab Switcher")
parser:flag("--prev", "Switch to previous window in history")
parser:flag("--notify", "Enable notifications (default)")
parser:flag("--no-notify", "Disable notifications")
parser:flag("--debug", "Enable debug logging")
parser:flag("--clear", "Clear saved state")
parser:flag("--apply", "Apply last selection")

local args = parser:parse()

local DEBUG = args["debug"] or false
local NOTIFY = args["notify"] or (not args["no-notify"])
local PREV = args["prev"] or false

local PREVIEW_NEXT = tonumber(os.getenv("HYPR_ALTAB_PREVIEW_NEXT")) or -1
local _capture_env = os.getenv("HYPR_ALTAB_CAPTURE")
local CAPTURE = (_capture_env == nil) or _capture_env == "1" or _capture_env == "true"

local xdg_runtime = os.getenv("XDG_RUNTIME_DIR") or "/tmp"
local his = os.getenv("HYPRLAND_INSTANCE_SIGNATURE") or ""
local socket_path = xdg_runtime .. "/hypr/" .. his .. "/.socket.sock"

local state_dir = xdg_runtime .. "/hypr-altab"
local state_file = state_dir .. "/state"
local preview_dir = state_dir .. "/previews"
local function log(msg)
    if DEBUG then
        io.stderr:write("[hypr-altab] " .. msg .. "\n")
    end
end
local function file_exists(path)
    local f = io.open(path, "r")
    if f then
        f:close()
        return true
    end
    return false
end

local function which(cmd)
    local h = io.popen("command -v '" .. cmd .. "' 2>/dev/null")
    local r = h:read("*a")
    h:close()
    return r and r:match("%S")
end
local function shell_escape(s)
    return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end
local function preview_path(addr)
    local safe = addr:gsub("0x", ""):gsub(":", "")
    return preview_dir .. "/" .. safe .. ".png"
end

local hyprctl_json

local function capture_preview(addr)
    if not CAPTURE then
        return
    end
    if not addr or addr == "" then
        return
    end
    if not which("grim") then
        return
    end
    os.execute("mkdir -p '" .. preview_dir .. "'")
    local path = preview_path(addr)

    local clients = hyprctl_json("j/clients")
    local match
    if type(clients) == "table" then
        for _, c in ipairs(clients) do
            if c.address == addr then
                match = c
                break
            end
        end
    end

    if match and type(match.at) == "table" and type(match.size) == "table" then
        local x = tonumber(match.at[1])
        local y = tonumber(match.at[2])
        local w = tonumber(match.size[1])
        local h = tonumber(match.size[2])
        if x and y and w and h and w > 1 and h > 1 then
            local geom = string.format("%d,%d %dx%d", x, y, w, h)
            os.execute("grim -g " .. shell_escape(geom) .. " " .. shell_escape(path) .. " >/dev/null 2>&1 &")
        end
    end
end
local function hyprctl_raw(cmd)
    if his == "" then
        log("HYPRLAND_INSTANCE_SIGNATURE not set")
        return ""
    end
    local ok, s = pcall(socket)
    if not ok or not s then
        log("socket init failed")
        return ""
    end
    local connected, err = s:connect(socket_path)
    if not connected then
        log("socket connect failed: " .. tostring(err))
        s:close()
        return ""
    end
    local sent, send_err = s:send(cmd)
    if not sent then
        log("socket send failed: " .. tostring(send_err))
        s:close()
        return ""
    end
    s:shutdown("send")
    local data, recv_err, partial = s:receive("*a")
    s:close()
    if not data then
        if recv_err and recv_err ~= "closed" then
            log("socket receive failed: " .. tostring(recv_err))
        end
        data = partial or ""
    end
    return data
end

hyprctl_json = function(cmd)
    local raw = hyprctl_raw(cmd)
    if not raw or raw == "" then
        return nil
    end
    local obj, _, err = json.decode(raw)
    return obj
end

local function contains_addr(list, addr)
    if type(list) ~= "table" or not addr then
        return false
    end
    for _, v in ipairs(list) do
        if v == addr then
            return true
        end
    end
    return false
end

local function build_notify(addr)
    local clients = hyprctl_json("j/clients")
    if type(clients) ~= "table" then
        return nil, nil, nil
    end
    local match
    for _, c in ipairs(clients) do
        if c.address == addr then
            match = c
            break
        end
    end
    if not match then
        return nil, nil, nil
    end
    local title = match.title or "(untitled)"
    local klass = match.class or "unknown"
    local ws = (match.workspace and match.workspace.name) or "?"
    local body = title .. "\n" .. klass .. "  •  " .. ws
    local icon = preview_path(addr)
    if file_exists(icon) then
        return klass, body, icon
    else
        return klass, body, nil
    end
end
local function notify_preview_pair(current_addr, next_addrs)
    if not which("notify-send") then
        return
    end
    local title, body, icon = build_notify(current_addr)
    if title then
        local cmd = "notify-send " .. shell_escape(title) .. " " .. shell_escape(body) .. " -t 2000 -r 6"
        if icon then
            cmd = cmd .. " -i " .. shell_escape(icon)
        end
        os.execute(cmd .. " &")
    end
end

local function read_state()
    local f = io.open(state_file, "r")
    if not f then
        return {}
    end
    local content = f:read("*a")
    f:close()
    local obj, _, err = json.decode(content)
    return obj or {}
end

local function write_state(tbl)
    os.execute("mkdir -p " .. state_dir)
    local f = io.open(state_file, "w")
    if not f then
        return
    end
    f:write(json.encode(tbl))
    f:close()
end

local function focus_addr(addr)
    if addr and addr ~= "" then
        hyprctl_raw("dispatch focuswindow address:" .. addr)
    end
end

local function build_history()
    local clients = hyprctl_json("j/clients")
    if type(clients) ~= "table" then
        return {}
    end
    local filtered = {}
    for _, c in ipairs(clients) do
        if c.mapped and (c.hidden == false or (c.grouped and #c.grouped > 0)) and type(c.focusHistoryID) == "number" then
            table.insert(filtered, c)
        end
    end
    table.sort(filtered, function(a, b)
        return a.focusHistoryID < b.focusHistoryID
    end)
    local addrs = {}
    for _, c in ipairs(filtered) do
        if c.address then
            table.insert(addrs, c.address)
        end
    end
    return addrs
end

local function main()
    local args = {}
    for i = 1, #arg do
        local flag = arg[i]:match("^%-%-(.+)$") or arg[i]
        args[flag] = true
    end
    if args["clear"] then
        os.remove(state_file)
        return
    end
    if args["apply"] then
        local state = read_state()
        local addr = state.addr
        log("apply: target=" .. tostring(addr))
        if addr then
            if not state.applied then
                focus_addr(addr)
                if CAPTURE then
                    capture_preview(addr)
                end
            end
        end
        os.remove(state_file)
        return
    end

    local state = read_state()
    local current_history = build_history()
    if #current_history < 2 then
        log("not enough windows in focus history")
        return
    end

    local history = current_history
    local state_age_ok = type(state.ts) == "number" and (os.time() - state.ts) <= 2
    if type(state.list) == "table" and #state.list > 1 and state_age_ok then
        local all_present = true
        for _, addr in ipairs(state.list) do
            if not contains_addr(current_history, addr) then
                all_present = false
                break
            end
        end
        if all_present then
            history = state.list
        end
    end

    local idx = tonumber(state.idx or 0)
    if PREV then
        idx = idx - 1
        if idx <= 0 then
            idx = #history - 1
        end
    else
        idx = idx + 1
        if idx >= #history then
            idx = 1
        end
    end

    local target = history[idx + 1]
    if not target then
        log("target missing after idx calc, rebuilding")
        history = current_history
        idx = PREV and (#history - 1) or 1
        target = history[idx + 1]
    end
    if not target then
        log("unable to resolve target window")
        return
    end

    local next_targets = {}
    local preview_count = (PREVIEW_NEXT > 0) and PREVIEW_NEXT or math.max(1, #history - 1)
    if #history > 1 and preview_count > 0 then
        for step = 1, preview_count do
            local next_idx
            if PREV then
                next_idx = idx - step
                if next_idx <= 0 then
                    next_idx = #history - 1 - ((step - idx) % math.max(1, #history - 1))
                end
            else
                next_idx = idx + step
                if next_idx >= #history then
                    next_idx = 1 + ((next_idx - #history) % math.max(1, #history - 1))
                end
            end
            table.insert(next_targets, history[next_idx + 1])
        end
    end
    log("preview: idx=" .. idx .. " target=" .. tostring(target) .. " count=" .. #history)
    write_state {
        idx = idx,
        addr = target,
        list = history,
        ts = os.time(),
        applied = true
    }
    focus_addr(target)
    if CAPTURE then
        capture_preview(target)
    end
    if NOTIFY then
        notify_preview_pair(target, next_targets)
    end
end

main()
